src/components/Camera.ts

```diff
@@ -0,0 +1,296 @@
+        };
+
+        this.availableDevices = [];
+        this.setCameraReference = this.setCameraReference.bind(this);
+        this.retakePicture = this.retakePicture.bind(this);
+        this.setStyle = this.setStyle.bind(this);
+        this.takePicture = this.takePicture.bind(this);
+        this.setUpWebCam = this.setUpWebCam.bind(this);
+        this.onClick = this.onClick.bind(this);
+        this.changeCamera = this.changeCamera.bind(this);
+        this.bindDimensions = this.bindDimensions.bind(this);
+    }
+
+    render() {
+        if (this.browserSupport === false) {
+            return this.renderAlert("This browser does not support the camera widget. Google-chrome is recommended.");
+        }
+        if (this.state.pictureTaken && this.state.screenshot) {
+            return this.renderPhoto();
+        }
+
+        return this.renderWebCam();
+    }
+
+    private renderWebCam(): ReactElement<{}> {
+        return createElement("div", { className: "widget-camera-parent", style: this.setStyle(this.props) },
+            createElement("div", {
+                ref: this.setCameraReference,
+                style: { filter: this.props.filter }
+            }),
+            this.createContainer({},
+                this.createButton(
+                    this.props.captureButtonName,
+                    this.props.captureButtonIcon,
+                    this.takePicture,
+                    "widget-camera-picture"
+                ), this.createSwitchCameraButton()
+            )
+        );
+    }
+
+    private renderPhoto(): ReactElement<{}> {
+        return createElement("div", { className: classNames("widget-camera-parent") },
+            createElement("img", {
+                alt: "Image could not be found!",
+                src: this.state.screenshot,
+                style: this.bindDimensions()
+            }),
+            this.createContainer(
+                { width: this.pictureWidth },
+                this.createButton(
+                    this.props.recaptureButtonName,
+                    this.props.captureButtonIcon,
+                    this.retakePicture,
+                    "widget-camera-picture"
+                ),
+                this.createButton(
+                    this.props.usePictureButtonName,
+                    this.props.usePictureButtonIcon,
+                    this.onClick,
+                    "widget-camera-switch-button"
+                )
+            )
+        );
+    }
+
+    private createIcon(buttonLabel: string, glyphiconName: string): ReactElement<{}> {
+        return (this.props.captionsToUse === "icons")
+            ? createElement("span", { className: classNames(`glyphicon glyphicon-${glyphiconName}`) })
+            : createElement("button", { className: classNames("btn btn-inverse active") }, buttonLabel);
+    }
+
+    private createButton(buttonLabel: string, glyphiconName: string, action: () => void, buttonclass: string): ReactElement<{}> {
```

Comment:
These create this component & create that component are all well & good, but they're not helping with the quantity of code in your component. They're also going to make testing a looot harder. Easier and better to just extract some of this into separate components.

Reply:
Done